R/group_apply.R

```diff
@@ -0,0 +1,133 @@
+#' This dataframe must contain the grouping variables and the necessary
+#' variables for the function you wish to call (see function-specific
+#' documentation for required columns).
+#' @param group \code{character}. A vector of column names, specifying the
+#' desired subgroups (e.g. "collection_no", "stage_bin"). Supplying more than
+#' one grouping variable will produce an output containing subgroups for each
+#' unique combination of values.
+#' @param fun \code{function}. The function you wish to apply to
+#' `occdf`. See details for compatible functions.
+#' @param ... Additional arguments available in the called
+#' function. These additional arguments may be required for function arguments
+#' without default values, or if you wish to overwrite the default argument
+#' value (see examples).
+#'
+#' @return A \code{dataframe} of the outputs from the selected function, with
+#' appended column(s) indicating the user-defined groups (`grouping`).
+#'
+#' @details `group_apply` applies functions to subgroups of data within a
+#' supplied dataset, enabling the separate analysis of occurrences or taxa from
+#' different time intervals, spatial regions, or trait values. The function
+#' serves as a wrapper around `palaeoverse` functions. Other functions which
+#' can be applied to a `data.frame` (e.g. `nrow`, `ncol`, `lengths`,
+#' `unique`) may also be used.
+#' \cr
+#' \cr
+#' All `palaeoverse` functions which require a dataframe input can be used in
+#' conjunction with the `group_apply` function. However, this is unnecessary
+#' for many functions (e.g. \code{\link{bin_time}}) as groups do not need to
+#' be partitioned before binning. This list provides
+#' users with `palaeoverse` functions that might be interesting to apply across
+#' group(s):
+#' - \code{\link{tax_unique}}: return the number of unique taxa per grouping
+#' variable.
+#' - \code{\link{tax_range_time}}: return the temporal range of taxa per
+#' grouping variable.
+#' - \code{\link{tax_range_geo}}: return the geographic range of taxa per
+#' grouping variable.
+#' - \code{\link{tax_check}}: return potential spelling variations of the
+#' same taxon per grouping variable. Note: `verbose` needs to be set to FALSE.
+#'
+#'
+#' @section Developer(s):
+#' Lewis A. Jones & William Gearty
+#' @section Reviewer(s):
+#' Kilian Eichenseer & Bethany Allen
+#' @importFrom stats as.formula
+#' @examples
+#' # Examples
+#' # Get tetrapods data
+#' occdf <- tetrapods
+#' # Remove NA data
+#' occdf <- subset(occdf, !is.na(genus))
+#' # Count number of occurrences from each country
+#' ex1 <- group_apply(occdf = occdf, group = "cc", fun = nrow)
+#' # Temporal range of taxa per country with default arguments
+#' ex2 <- group_apply(occdf = occdf, group = "cc", fun = tax_range_time)
+#' # Temporal range of taxa per country with updated arguments
+#' ex3 <- group_apply(occdf = occdf,
+#'                    group = c("cc"),
+#'                    fun = tax_range_time,
```

Comment:
I think we should update `tax_range_time` to `tax_unique` in the examples. This will be more common usage?

Reply:
I don't have any issue with the examples using `tax_range_time`, but we should probably include an example with `tax_unique` (and/or add an example to `tax_unique` that uses `group_apply`)

Reply:
Adding this to my issue log to integrate next week!